src/main/scala/uoption/package.scala

```diff
@@ -0,0 +1,63 @@
+import scala.language.higherKinds
+
+package object uoption {
+  private[uoption] sealed case class WrappedNone(depth: Int)(
+      val unwrap: Any // UNone or WrappedNone
+  ) {
+    lazy val wrap: WrappedNone =
+      WrappedNone(depth + 1)(this)
+
+    private val stringRepr: String =
+      ("USome(" * depth) + "UNone" + (")" * depth)
+
+    override def toString(): String = stringRepr
+  }
+
+  object UNone {
+    private[uoption] val wrap: WrappedNone = WrappedNone(1)(this)
+  }
+
+  type UOption[+A] >: UNone.type <: AnyRef
+
+  type USome[+A] <: UOption[A]
+
+  object USome {
+    @inline // only for Scala.js?
+    def apply[A](value: A): USome[A] = value match {
+      case value @ UNone      => value.wrap.asInstanceOf[USome[A]]
+      case value: WrappedNone => value.wrap.asInstanceOf[USome[A]]
+      case _                  => value.asInstanceOf[USome[A]]
+    }
+
+    // UOptionOps fits the contract of name-based pattern matching
+    def unapply[A](option: UOption[A]): UOptionOps[A] =
+      new UOptionOps(option)
+  }
+
+  implicit class UOptionOps[A](val __self: UOption[A]) extends AnyVal {
```

Comment:
Ah good point. It was initially written to support 2.10, but I had to remove 2.10 for other reasons so we can fix that one too, indeed!